src/sdk/pynni/tests/test_graph_utils.py

```diff
@@ -0,0 +1,111 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+import sys
+import os
+import math
+import uuid
+import shutil
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from tensorboard.compat.proto.graph_pb2 import GraphDef
+from google.protobuf import text_format
+from unittest import TestCase, main
+
+from nni.graph_utils import build_module_graph, build_graph
+
+class BackboneModel1(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv1 = nn.Conv2d(1, 1, 1, 1)
+    def forward(self, x):
+        return self.conv1(x)
+
+class BackboneModel2(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.conv1 = nn.Conv2d(1, 20, 5, 1)
+        self.conv2 = nn.Conv2d(20, 50, 5, 1)
+        self.bn1 = nn.BatchNorm2d(self.conv1.out_channels)
+        self.bn2 = nn.BatchNorm2d(self.conv2.out_channels)
+        self.fc1 = nn.Linear(4 * 4 * 50, 500)
+        self.fc2 = nn.Linear(500, 10)
+
+    def forward(self, x):
+        x = F.relu(self.bn1(self.conv1(x)))
+        x = F.max_pool2d(x, 2, 2)
+        x = F.relu(self.bn2(self.conv2(x)))
+        x = F.max_pool2d(x, 2, 2)
+        x = x.view(x.size(0), -1)
+        
+        x = F.relu(self.fc1(x))
+        x = self.fc2(x)
+        return x
+
+class BigModel(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.backbone1 = BackboneModel1()
+        self.backbone2 = BackboneModel2()
+        self.fc3 = nn.Linear(10, 2) 
+    def forward(self, x):
+        x = self.backbone1(x)
+        x = self.backbone2(x)
+        x = self.fc3(x)
+        return x
```

Comment:
I'd like two more models to be added as the graph utils provided by PyTorch fail for them.

```python
class MyModule(nn.Module):
  def __init__(self):
    super().__init__()
    self.weight = nn.Linear(5, 3)
    self.bias = nn.Linear(5, 3)
    self.module = nn.Linear(6, 1)

  def forward(self, x):
    tensors = [self.weight(x), self.bias(x)]
    self.module(torch.cat(tensors, dim=1))
    return x
```

```python
class MyModule(nn.Module):
  def __init__(self):
    super().__init__()
    self.module = nn.ModuleList([
      nn.Linear(5, 3),
      nn.Linear(3, 1)
    ])

  def forward(self, x):
    x = self.module[0](x)
    x = self.module[1](x)
    return x
```

Reply:
ok, can I save the output of your code as expected file?

Reply:
Sure.

First one:

Code:

```python
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.weight = nn.Linear(5, 3)
        self.bias = nn.Linear(5, 3)
        self.module = nn.Linear(6, 1)

    def forward(self, x):
        tensors = [self.weight(x), self.bias(x)]
        self.module(torch.cat(tensors, dim=1))
        return x
```

Expected result:

```
node {
  name: "input/input.1"
  op: "IO Node"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "output/output.1"
  op: "IO Node"
  input: "input/input.1"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "MyModule/Linear[weight]/bias/49"
  op: "prim::GetAttr"
  input: "MyModule/Linear[weight]/weight/35"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/weight/50"
  op: "prim::GetAttr"
  input: "MyModule/Linear[weight]/weight/35"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/51"
  op: "aten::t"
  input: "MyModule/Linear[weight]/weight/50"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/52"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/53"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[weight]/54"
  op: "aten::addmm"
  input: "MyModule/Linear[weight]/bias/49"
  input: "input/input.1"
  input: "MyModule/Linear[weight]/51"
  input: "MyModule/Linear[weight]/52"
  input: "MyModule/Linear[weight]/53"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/bias/55"
  op: "prim::GetAttr"
  input: "MyModule/Linear[bias]/weight/38"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/weight/56"
  op: "prim::GetAttr"
  input: "MyModule/Linear[bias]/weight/38"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/57"
  op: "aten::t"
  input: "MyModule/Linear[bias]/weight/56"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/58"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/59"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/Linear[bias]/60"
  op: "aten::addmm"
  input: "MyModule/Linear[bias]/bias/55"
  input: "input/input.1"
  input: "MyModule/Linear[bias]/57"
  input: "MyModule/Linear[bias]/58"
  input: "MyModule/Linear[bias]/59"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/23"
  op: "prim::ListConstruct"
  input: "MyModule/Linear[weight]/54"
  input: "MyModule/Linear[bias]/60"
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/24"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/input"
  op: "aten::cat"
  input: "MyModule/23"
  input: "MyModule/24"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 6
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/61"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
versions {
  producer: 22
}
```

Reply:
Second one:

Code:

```python
class MyModule(nn.Module):
    def __init__(self):
        super().__init__()
        self.module = nn.ModuleList([
            nn.Linear(5, 3),
            nn.Linear(3, 1)
        ])

    def forward(self, x):
        x = self.module[0](x)
        x = self.module[1](x)
        return x
```

Expected result:

```
node {
  name: "input/input.1"
  op: "IO Node"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 5
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "output/output.1"
  op: "IO Node"
  input: "MyModule/ModuleList[module]/Linear[1]/46"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: ""
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/bias/35"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/26"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/weight/36"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/26"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/37"
  op: "aten::t"
  input: "MyModule/ModuleList[module]/Linear[0]/weight/36"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 5
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/38"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/39"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[0]/input"
  op: "aten::addmm"
  input: "MyModule/ModuleList[module]/Linear[0]/bias/35"
  input: "input/input.1"
  input: "MyModule/ModuleList[module]/Linear[0]/37"
  input: "MyModule/ModuleList[module]/Linear[0]/38"
  input: "MyModule/ModuleList[module]/Linear[0]/39"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 3
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/bias/41"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/30"
  attr {
    key: "attr"
    value {
      s: "{ name :  bias }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/weight/42"
  op: "prim::GetAttr"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/30"
  attr {
    key: "attr"
    value {
      s: "{ name :  weight }"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/43"
  op: "aten::t"
  input: "MyModule/ModuleList[module]/Linear[1]/weight/42"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 3
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/44"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/45"
  op: "prim::Constant"
  attr {
    key: "attr"
    value {
      s: "{ value : 1}"
    }
  }
}
node {
  name: "MyModule/ModuleList[module]/Linear[1]/46"
  op: "aten::addmm"
  input: "MyModule/ModuleList[module]/Linear[1]/bias/41"
  input: "MyModule/ModuleList[module]/Linear[0]/input"
  input: "MyModule/ModuleList[module]/Linear[1]/43"
  input: "MyModule/ModuleList[module]/Linear[1]/44"
  input: "MyModule/ModuleList[module]/Linear[1]/45"
  attr {
    key: "_output_shapes"
    value {
      list {
        shape {
          dim {
            size: 4
          }
          dim {
            size: 1
          }
        }
      }
    }
  }
  attr {
    key: "attr"
    value {
      s: "{}"
    }
  }
}
versions {
  producer: 22
}
```

Reply:
Test them with:

```python
model = MyModule()
graph_def, _ = graph(model, torch.randn(4, 5), verbose=False)
result = text_format.MessageToString(graph_def)

print(result)
```